File: src/csharp/pulse/cdm/properties/SEScalarMassPerTime.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class MassPerTimeUnit : Unit
  {
    private MassPerTimeUnit(string v) : base(v) { }

    public static MassPerTimeUnit g_Per_s = new MassPerTimeUnit("g/s");
    public static MassPerTimeUnit g_Per_min = new MassPerTimeUnit("g/min");
    public static MassPerTimeUnit g_Per_day = new MassPerTimeUnit("g/day");
    public static MassPerTimeUnit mg_Per_s = new MassPerTimeUnit("mg/s");
    public static MassPerTimeUnit mg_Per_min = new MassPerTimeUnit("mg/min");
    public static MassPerTimeUnit ug_Per_s = new MassPerTimeUnit("ug/s");
    public static MassPerTimeUnit kg_Per_s = new MassPerTimeUnit("kg/s");
    public static MassPerTimeUnit ug_Per_min = new MassPerTimeUnit("ug/min");

    public static MassPerTimeUnit FromString(string u)
    {
      if (u == MassPerTimeUnit.g_Per_s.ToString())
        return MassPerTimeUnit.g_Per_s;
      if (u == MassPerTimeUnit.g_Per_min.ToString())
        return MassPerTimeUnit.g_Per_min;
      if (u == MassPerTimeUnit.g_Per_day.ToString())
        return MassPerTimeUnit.g_Per_day;
      if (u == MassPerTimeUnit.mg_Per_s.ToString())
        return MassPerTimeUnit.mg_Per_s;
      if (u == MassPerTimeUnit.mg_Per_min.ToString())
        return MassPerTimeUnit.mg_Per_min;
      if (u == MassPerTimeUnit.ug_Per_s.ToString())
        return MassPerTimeUnit.ug_Per_s;
      if (u == MassPerTimeUnit.kg_Per_s.ToString())
        return MassPerTimeUnit.kg_Per_s;
      if (u == MassPerTimeUnit.ug_Per_min.ToString())
        return MassPerTimeUnit.ug_Per_min;

      throw new System.ArgumentException(u + " is not a know mass per time unit");
    }

  }

  public class SEScalarMassPerTime : SEScalar
  {
    protected MassPerTimeUnit unit;

    public SEScalarMassPerTime() : base()
    {

    }

    public SEScalarMassPerTime(SEScalarMassPerTime from)
    {
      Set(from);
    }

    public SEScalarMassPerTime(double value, MassPerTimeUnit unit)
    {
      SetValue(value, unit);
    }

    public void Set(SEScalarMassPerTime from)
    {
      if (from == null)
      {
        Invalidate();
        return;
      }
      SetValue(from.value, from.unit);
    }

    public new void SetValue(double value)
    {
      throw new System.NotSupportedException("You must provide a unit");
    }

    public void SetValue(double value, MassPerTimeUnit unit)
    {
      this.value = value;
      this.unit = unit;
    }

    public double GetValue(MassPerTimeUnit unit)
    {
      return UnitConverter.Convert(this.value, this.unit, unit);
    }

    public MassPerTimeUnit GetUnit()
    {
      return unit;
    }

    public new string ToString()
    {
      return this.value + "(" + this.unit.ToString() + ")";
    }
  }
}

File: src/cpp/cdm/properties/SEScalarMassPerTime.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/properties/SEScalarMassPerTime.h"

const MassPerTimeUnit MassPerTimeUnit::g_Per_s("g/s");
const MassPerTimeUnit MassPerTimeUnit::g_Per_min("g/min");
const MassPerTimeUnit MassPerTimeUnit::g_Per_day("g/day");
const MassPerTimeUnit MassPerTimeUnit::mg_Per_s("mg/s");
const MassPerTimeUnit MassPerTimeUnit::mg_Per_min("mg/min");
const MassPerTimeUnit MassPerTimeUnit::ug_Per_s("ug/s");
const MassPerTimeUnit MassPerTimeUnit::kg_Per_s("kg/s");
const MassPerTimeUnit MassPerTimeUnit::ug_Per_min("ug/min");

bool MassPerTimeUnit::IsValidUnit(const std::string& unit)
{
  if (g_Per_s.GetString().compare(unit) == 0)
    return true;
  if (g_Per_min.GetString().compare(unit) == 0)
    return true;
  if (g_Per_day.GetString().compare(unit) == 0)
    return true;
  if (mg_Per_s.GetString().compare(unit) == 0)
    return true;
  if (mg_Per_min.GetString().compare(unit) == 0)
    return true;
  if (ug_Per_s.GetString().compare(unit) == 0)
    return true;
  if (kg_Per_s.GetString().compare(unit) == 0)
    return true;
  if (ug_Per_min.GetString().compare(unit) == 0)
    return true;
  return false;
}

const MassPerTimeUnit& MassPerTimeUnit::GetCompoundUnit(const std::string& unit)
{
  if (g_Per_s.GetString().compare(unit) == 0)
    return g_Per_s;
  if (g_Per_min.GetString().compare(unit) == 0)
    return g_Per_min;
  if (g_Per_day.GetString().compare(unit) == 0)
    return g_Per_day;
  if (mg_Per_s.GetString().compare(unit) == 0)
    return mg_Per_s;
  if (mg_Per_min.GetString().compare(unit) == 0)
    return mg_Per_min;
  if (ug_Per_s.GetString().compare(unit) == 0)
    return ug_Per_s;
  if (kg_Per_s.GetString().compare(unit) == 0)
    return kg_Per_s;
  if (ug_Per_min.GetString().compare(unit) == 0)
    return ug_Per_min;
  throw CommonDataModelException(unit + " is not a valid MassPerTime unit");
}

template class SEScalarQuantity<MassPerTimeUnit>;
